examples/nginx-connect-proxy.md

```diff
@@ -0,0 +1,254 @@
+
+To connect to the mesh network, we need the required TLS certificates. We'll
+need the root certificate (the CA) and the client certificates. We will use a
+separate template for each certificate, to put each in their own file.
+
+#### ca.crt
+```liquid
+cat > ca.crt.tmpl << EOF
+{{range caRoots}}{{.RootCertPEM}}{{end}}
+EOF
+```
+#### cert.pem
+```liquid
+cat > cert.pem.tmpl << EOF
+{{with caLeaf "ingress"}}{{.CertPEM}}{{end}}
+EOF
+```
+#### cert.key
+```liquid
+cat > cert.key.tmpl << EOF
+{{with caLeaf "ingress"}}{{.PrivateKeyPEM}}{{end}}
+EOF
+```
+
+With all the certificate templates in place, we now just need the configuration
+file template for the NGINX proxy. It uses the `connect` template function to
+get all the "webserver" connect enabled services.
+
+#### nginx-proxy.conf.tmpl
+```nginx
+cat > nginx-proxy.conf.tmpl << EOF
+daemon off;
+master_process off;
+pid nginx.pid;
+error_log /dev/stdout;
+
+events {}
+
+http {
+  access_log /dev/stdout;
+
+  server {
+    listen 8080 default_server;
+
+    location / {
+{{range connect "webserver"}}
+      proxy_pass https://{{.Address}}:{{.Port}};
+{{end}}
+      # these refer to files written by templates above
+      proxy_ssl_certificate cert.pem;
+      proxy_ssl_certificate_key cert.key;
+      proxy_ssl_trusted_certificate ca.crt;
+    }
+  }
+}
+EOF
+```
+
+## Running the NGINX Ingress Proxy with Consul-Template
+
+Using the consul-template configuration below, we run the NGINX proxy.
+
+```shell
+$ consul-template -config ingress-config.hcl -log-level=info
+```
+
+Consul-template will restart the nginx server when any of the templates are
+rendered (files written). Note that if multiple templates are rendered, it will
+still only restart it once.
+
+#### ingress-config.hcl
+```hcl
+cat > ingress-config.hcl << EOF
+exec {
+  command = "/usr/sbin/nginx -p . -c nginx-proxy.conf"
+}
+template {
+  source = "ca.crt.tmpl"
+  destination = "ca.crt"
+}
+template {
+  source = "cert.pem.tmpl"
+  destination = "cert.pem"
+}
+template {
+  source = "cert.key.tmpl"
+  destination = "cert.key"
+}
+template {
+  source = "nginx-proxy.conf.tmpl"
+  destination = "nginx-proxy.conf"
+}
+EOF
+```
+
+## Test it
+```shell
+$ curl http://localhost:8080
+Welcome to nginx!
+```
+
+## Intentions Work
+
+Because the ingress service was registered with consul, you can control what
+the it is able to connect to in the mesh network with [intentions][intentions].
+
+```shell
+$ curl http://localhost:8080
+Welcome to nginx!
+$ consul intention create -deny  ingress webserver
+Created: ingress => webserver (deny)
+$ curl http://localhost:8080
+<html>
+<head><title>502 Bad Gateway</title></head>
+<body bgcolor="white">
+<center><h1>502 Bad Gateway</h1></center>
+<hr><center>nginx/1.14.2</center>
+</body>
+</html>
+$ consul intention delete ingress webserver
+Intention deleted.
+$ curl http://localhost:8080
+Welcome to nginx!
+```
+
+## Using the shell script
+
```

Comment:
Maybe add a sentence here that you can automate the process using the script in the repo.

Reply:
Added... `Included is a shell script that will extract the config files from this document and runs all the different services in multiple [GNU Screen][screen] windows. The script provides some basic instructions.`

I'm adding this to the intro-output of the script... ```To switch between the different screen windows type control-a then the number
0-4 (eg. 'ctrl-a 0' will take you to the first window, the web-server). When
done just exit each window (type 'exit' or ctrl-d), exiting the last will
exit screen.```